Managed clusters

antiguru opened the pull request:
### Motivation

Implement managed clusters as outlined in #19554

This PR is a first step towards offering managed clusters. It
* implements support for creating and dropping managed clusters,
* offers support for altering managed clusters to set and reset attributes,
* implement support for migrating managed to non-managed clusters and vice versa, with some limitations.

The PR at this point does not
* provide graceful configuration application, i.e., changing the size will tear down all replicas and create new ones in the same operation.

Open questions
* Should the availability zone really be availability zone`s`? This wouldn't change the implementation dramatically, but we should clarify it now.
* The replica names are not `mz_rX`, but `managed_rX` because otherwise dropping objects is hard. This can be fixed with moving validation elsewhere, but seems to be out-of-scope for this first step. (The reason is that we're executing generic `DropObject` operations, so we don't know whether dropping something with a reserved name is OK or not.)

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

jkosh44 commented on src/adapter/src/catalog.rs:
```diff
@@ -6322,6 +6334,7 @@ impl Catalog {
                     replica_id,
                     name,
                     to_name,
+                    user_initiated,
```
I thought this wasn't needed anymore?

antiguru replied:
Removed.

jkosh44 commented on src/adapter/src/catalog.rs:
```diff
@@ -6603,6 +6623,12 @@ impl Catalog {
                     ObjectId::Role(_) => unreachable!("roles have no owner"),
                 },
+                Op::UpdateClusterConfig { id, config } => {
+                    let cluster = state.get_cluster_mut(id);
+                    cluster.config = config;
+                    tx.update_cluster(id, cluster)?;
+                    info!("update cluster {}", cluster.name);
```
Do we store the config in any builtin table anywhere? If so, we'll have to update that table as well.

antiguru replied:
Done.

jkosh44 commented on src/stash/protos/objects.proto:
```diff
@@ -64,6 +64,7 @@ message ClusterValue {
     GlobalId linked_object_id = 2;
     RoleId owner_id = 3;
     repeated MzAclItem privileges = 4;
+    ClusterConfig config = 5;
```
This probably requires a stash migration to migrate the existing clusters.

antiguru replied:
I'm not sure! In https://github.com/MaterializeInc/materialize/pull/19762/files#diff-ae4537b9c6964f950b962c46de8fa54ca0a55bd2328725b122e6dc2e4d722c3bR271 I default to `Unmanaged` if the proto field isn't set, which should match the current behavior. Let's look into this once the PR is a bit further.

ParkMyCar replied:
Defaulting to `Unmanaged` in this way should be fine. There are maybe two small issues I can see though:

1. Someone changes the default value, which results in external changes even though the serialized data didn't change.
2. Somewhere else (e.g. a migration) we don't consider `None == Unmanaged`.

I don't think either of these are too big of a risk though, so I'm fine with this

antiguru replied:
Added a stash migration.

jkosh44 commented on src/adapter/src/catalog/storage/stash.rs:
```diff
@@ -595,6 +596,26 @@ pub async fn initialize(
+pub fn default_cluster_config(_args: &BootstrapArgs) -> proto::ClusterConfig {
+    proto::ClusterConfig {
+        variant: Some(proto::cluster_config::Variant::Unmanaged(proto::Empty {})),
+        // TODO: Should the default cluster be managed or unmanaged?
+        // variant: Some(proto::cluster_config::Variant::Managed(
+        //     proto::cluster_config::ManagedCluster {
+        //         size: args.default_cluster_replica_size.to_string(),
+        //         availability_zone: Some(args.default_availability_zone.to_string()),
+        //         logging: Some(proto::ReplicaLogging {
+        //             log_logging: false,
+        //             interval: Some(proto::Duration::from_secs(1)),
+        //         }),
+        //         idle_arrangement_merge_effort: None,
+        //         replication_factor: 1,
+        //     },
+        // )),
```
My opinion on this is that if we're going to try and push everyone to start using managed clusters over unmanaged clusters, then we should make the default cluster managed. If we decide on managed clusters, then a similar question is, should we alter existing default clusters to be managed?

antiguru replied:
Filed #19976 to follow up.

jkosh44 commented on src/adapter/src/coord/sequencer/cluster.rs:
I think we should do this refactor in a separate PR or separate commit.

antiguru replied:
Done.

jkosh44 reviewed (commented):
Left some high level comments. Overall the structure looks good, but I did skim some of the more denser parts.

antiguru commented on src/stash/protos/hashes.json:
```diff
@@ -1,7 +1,7 @@
   {
     "name": "objects.proto",
-    "md5": "0c3e0fc9477124bbb7ea01902e8f999f"
+    "md5": "f085f2648f065143e0c9b12b572435c7"
```
Thanks, will do once the PR gets closer to a finished version.

antiguru replied:
Done.

philip-stoev commented:
Item No 1 . We allow REPLICATION FACTOR 0:

```
materialize=> CREATE CLUSTER foo SIZE '1', replication factor 0;
CREATE CLUSTER
```

philip-stoev commented:
Item No 2. We need to prevent large replication factors from being used. This way we can save users from fat-finger mistakes in cases where they have a large max_replicas_per_cluster limit:
```
materialize=> CREATE CLUSTER foo SIZE '1', replication factor 99;
ERROR:  creating cluster replica would violate max_replicas_per_cluster limit (desired: 99, limit: 5, current: 0)
HINT:  Drop an existing cluster replica or contact sales to request a limit increase.
```

philip-stoev commented:
Item No 3. (Related but separate) The existing `max_replicas_per_cluster` should be enforced at `CREATE` or `ALTER` time of the entire cluster. Currently, the `max_replicas_per_cluster` kicks in only after a bunch of replicas have been created . Instead, it should have failed immediately:

```
materialize=> CREATE CLUSTER foo SIZE '1', replication factor 999;
ERROR:  creating cluster replica would violate max_replicas_per_cluster limit (desired: 999, limit: 5, current: 0)
HINT:  Drop an existing cluster replica or contact sales to request a limit increase.
Time: 30044.206 ms (00:30.044)
```

philip-stoev commented:
Item No 4. The check whether a cluster exists or not seems misplaced in the code, or something else fishy is going on:

```
materialize=> CREATE CLUSTER foo SIZE '1', replication factor 1;
CREATE CLUSTER
Time: 155.702 ms

materialize=> CREATE CLUSTER foo SIZE '1', replication factor 99;
ERROR:  creating cluster replica would violate max_replicas_per_cluster limit (desired: 99, limit: 5, current: 0)
HINT:  Drop an existing cluster replica or contact sales to request a limit increase.
Time: 3815.832 ms (00:03.816)
````
 This should have failed immediately with "cluster already exists", without any other work being done. Instead, there is a period of 150% CRDB CPU usage that is unexpected.

philip-stoev reviewed (approved):
I spent a bit of time manually testing today. I think this PR needs the following additional tests:

- a Platform Checks test to cover upgrades and restarts
- a dedicated test/cloudtest or a test/testdrive/.td test that exercises the new functionality in the context of the K8s process orchestrator.

Unfortunately I am unable to create those tests right now as I will be OOO , but you are welcome to give it a try (@def- will be around to help) or open a follow-up ticket for the QA team.

Thank you.

jkosh44 commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -3989,6 +4071,134 @@ pub fn plan_alter_index_options(
+            match managed.unwrap_or_else(|| cluster.is_managed()) {
+                true => {
+                    if replica_defs.is_some() {
+                        sql_bail!("REPLICAS not supported for managed clusters");
```
Technically yes, but `sql_bail!` is so prevalent that it's not the end of the world. If you can find better structured errors or don't mind creating some new ones, then those would be preferred.

antiguru replied:
#20000 to track this work.

def- commented on doc/user/content/get-started/key-concepts.md:
```diff
@@ -183,6 +183,12 @@ meaning that on a given physical replica, objects in the cluster are in
+
+Materialize offers managed clusters to automatically manage the replicas associated with a cluster.
+You specify a configuration, such as the size of replicas and the replication factor, and
+Materialize will manage the replicas and apply changes are required.
```
```suggestion
Materialize will manage the replicas and apply changes as required.
```

def- reviewed (changes requested):
I'll take a look from QA side and implement the tests Philip proposed. If you don't want to wait, feel free to submit before

def- commented on doc/user/content/sql/create-cluster.md:
```diff
@@ -55,6 +55,21 @@ Clusters containing sources and sinks can have at most one replica.
+## Managed and unmanaged clusters
+
+Materialize offers two variants of cluster, managed and unmanaged clusters.
+While both can be interchangeably used to host computations, the differ in how they are configured.
```
```suggestion
While both can be interchangeably used to host computations, they differ in how they are configured.
```

def- commented on src/stash/src/upgrade/v18_to_v19.rs:
```diff
@@ -0,0 +1,11 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+/// No-op migration. Just need to bump the version number after supporting cluster/replica renaming.
```
I think it's not about renaming, but about SET/RESET.

def- commented:
There could be an argument that the `size` should be checked for validity immediately, and not just when it has an effect:
```
materialize=> CREATE CLUSTER foo SIZE 'bar', replication factor 0;
CREATE CLUSTER
materialize=> alter cluster foo reset (replication factor);
ERROR:  unknown cluster replica size bar
HINT:  Valid cluster replica sizes are: 1, 16, 2, 32, 4, 8, mem-16, mem-2, mem-32, mem-4, mem-8, 2-1, 2-2, 2-4, 4-1, 4-4, 8-1, 8-8, 16-1, 16-16, 32-1, 32-32
```

ParkMyCar commented on src/stash/protos/objects_v19.proto:
Thanks for the ping, I just put up #19921 which includes a test to guard against this

def- commented:
@antiguru I asked @nrainer-materialize for some help about the cloudtest failure, he's currently looking into cloudtest anyway.

nrainer-materialize commented:
Hi @antiguru, as promised I had a look at the failing cloud test `test_managed_cluster_sizing`.

### Summary
The problem is that after running `ALTER CLUSTER sized1 SET (REPLICATION FACTOR 4)` not all pods reach the state `Running` but a few remain stuck in `Pending`.

### More details
Information is taken from a local execution of the test case.

Relevant pods at the beginning of the test case are:

```
NAME                                READY   STATUS      RESTARTS   AGE
cluster-s1-replica-2-0              1/1     Running     0          10s
cluster-s2-replica-3-0              1/1     Running     0          10s
cluster-u1-replica-1-0              1/1     Running     0          10s
```

After executing `CREATE CLUSTER sized1 SIZE '2-1', REPLICATION FACTOR 2")`, they are:

```
cluster-s1-replica-2-0              1/1     Running     0          104s
cluster-s2-replica-3-0              1/1     Running     0          104s
cluster-u1-replica-1-0              1/1     Running     0          104s
---
cluster-u2-replica-4-0              1/1     Running     0          13s
cluster-u2-replica-4-1              1/1     Running     0          13s
cluster-u2-replica-5-0              1/1     Running     0          13s
cluster-u2-replica-5-1              1/1     Running     0          13s
```

Then, after executing `ALTER CLUSTER sized1 SET (REPLICATION FACTOR 4)`, they are:

```
NAME                                READY   STATUS      RESTARTS   AGE
cluster-s1-replica-2-0              1/1     Running     0          3m9s
cluster-s2-replica-3-0              1/1     Running     0          3m9s
cluster-u1-replica-1-0              1/1     Running     0          3m9s
---
cluster-u2-replica-4-0              1/1     Running     0          98s
cluster-u2-replica-4-1              1/1     Running     0          98s
cluster-u2-replica-5-0              1/1     Running     0          98s
cluster-u2-replica-5-1              1/1     Running     0          98s
---
cluster-u2-replica-6-0              1/1     Running     0          7s
cluster-u2-replica-6-1              1/1     Running     0          7s
cluster-u2-replica-7-0              0/1     Pending     0          7s             <= remains pending!
cluster-u2-replica-7-1              0/1     Pending     0          7s             <= remains pending!
```

Four new pods get started after increasing the replication factor from `2` to `4`. However, `cluster-u2-replica-7-0` and `cluster-u2-replica-7-1` remain (indefinitely) in state `Pending`. Consequently, the test case fails because of a timeout when `waiting for kubectl wait --for condition=Ready pod/cluster-u2-replica-7-0 --timeout 300s --context kind-cloudtest` is executed due to the pod not getting ready.

### Other
* This test case was newly introduced on your branch.

I hope this helps. Let me know if you need further information.

jkosh44 commented on src/adapter/src/catalog.rs:
```diff
@@ -6603,6 +6615,14 @@ impl Catalog {
                     }
                     ObjectId::Role(_) => unreachable!("roles have no owner"),
                 },
+                Op::UpdateClusterConfig { id, name, config } => {
+                    builtin_table_updates.push(state.pack_cluster_update(&name, -1));
+                    let cluster = state.get_cluster_mut(id);
+                    cluster.config = config;
+                    tx.update_cluster(id, cluster)?;
+                    builtin_table_updates.push(state.pack_cluster_update(&name, 1));
+                    info!("update cluster {}", name);
+                }
```
There doesn't seem to be a specific rule (that I'm aware of) about what goes in the audit log. This seems like it would be useful to have in the audit log though, so I would recommend adding it.

antiguru replied:
Filed #19978 to implement this.

antiguru commented:
@nrainer-materialize  thanks for the explanation! I think the cloudtest consumes more resources than our kind setup offers:
```
| Events:                                                                                    │
│   Type     Reason            Age    From               Message                             │
│   ----     ------            ----   ----               -------                             │
│   Warning  FailedScheduling  4m15s  default-scheduler  0/4 nodes are available: 1 node(s)  │
│ didn't match pod anti-affinity rules, 1 node(s) had untolerated taint {node-role.kubernete │
│ s.io/control-plane: }, 2 node(s) didn't match Pod's node affinity/selector. preemption: 0/ │
│ 4 nodes are available: 1 No preemption victims found for incoming pod, 3 Preemption is not │
│  helpful for scheduling..                                                                  │
```

@def- Could it be the test you added wants too many replicas?

def- commented:
Sorry, was not aware of that limit for cloudtest, reduced it now.

antiguru commented on src/adapter/src/rbac.rs:
```diff
@@ -1116,6 +1121,7 @@ fn generate_required_privileges(
         }
         Plan::AlterClusterRename(AlterClusterRenamePlan { .. })
         | Plan::AlterClusterReplicaRename(AlterClusterReplicaRenamePlan { .. })
+        | Plan::AlterClusterSetOptions(AlterClusterSetOptionsPlan { .. })
```
Fixed.

antiguru commented:
> Item No 1 . We allow REPLICATION FACTOR 0:

That's intended to allow a cluster without replicas.

> Item No 2. We need to prevent large replication factors from being used. This way we can save users from fat-finger mistakes in cases where they have a large max_replicas_per_cluster limit:

This should be part of the op validation logic. Unfortunately, it needs to see the create/drop replica ops to determine whether limits are impacted or not. This happens before executing any op, so it should not cause side-effects as far as I understand. Maybe @jkosh44 can clarify this?

> Item No 3. (Related but separate) The existing max_replicas_per_cluster should be enforced at CREATE or ALTER time of the entire cluster. Currently, the max_replicas_per_cluster kicks in only after a bunch of replicas have been created . Instead, it should have failed immediately:

Idem.

> Item No 4. The check whether a cluster exists or not seems misplaced in the code, or something else fishy is going on:

Fixed.

> Item No 5. I think this can use the default cluster size setting rather than throwing an error:

I don't want to introduce a default cluster size as part of this PR, but would prefer to do this in follow-up work.

> Item No 6. One is allowed to increase the replication factor of a storage cluster to more than 1. I think it is unsafe:

Fixed. Managed clusters cannot be used to back storage, and migrations check that a cluster is not linked.

> So we expect the replicas to follow a certain naming convention so that they can be converted to managed?

Yes. It's something that we reserve the right to change in the future, and I'm wondering whether we should simply use the `r1`, `r2`, ... pattern.

> Item No 7. Not sure if this is within the scope of this PR, but resizing the replicas does not seem to happen in a way that ensures there is no downtime:

Known limitation, future work to add a smarter policy.

> There could be an argument that the size should be checked for validity immediately, and not just when it has an effect:

Fixed.

> An empty unmanaged cluster also can't be converted to a managed, while managed can be created empty:

We don't have a default size, which is why we cannot migrate an empty unmanaged cluster.

jkosh44 commented:
> > Item No 2. We need to prevent large replication factors from being used. This way we can save users from fat-finger mistakes in cases where they have a large max_replicas_per_cluster limit:
> 
> This should be part of the op validation logic. Unfortunately, it needs to see the create/drop replica ops to determine whether limits are impacted or not. This happens before executing any op, so it should not cause side-effects as far as I understand. Maybe @jkosh44 can clarify this?

I don't think I fully understand the concern here. Is the error message
```
materialize=> CREATE CLUSTER foo SIZE '1', replication factor 99;
ERROR:  creating cluster replica would violate max_replicas_per_cluster limit (desired: 99, limit: 5, current: 0)
HINT:  Drop an existing cluster replica or contact sales to request a limit increase.
```
part of the current behavior for the PR? If so it does seem to prevent large replication factors from being used and does saves users from fat-finger mistakes. If the `max_replicas_per_cluster` limit is 99, then we have to allow users to make a cluster with 99 replicas otherwise the limit would be meaningless.

jkosh44 commented:
>> Item No 3. (Related but separate) The existing max_replicas_per_cluster should be enforced at CREATE or ALTER time of the entire cluster. Currently, the max_replicas_per_cluster kicks in only after a bunch of replicas have been created . Instead, it should have failed immediately:
>
> Idem.

@philip-stoev Are you sure about this? Would you mind providing an example to show this. All resource limit checks should occur before any resources have actually been creating. Failing the resource limit check should prevent any resource from being created.

antiguru commented on src/adapter/src/coord/sequencer/cluster.rs:
```diff
@@ -42,32 +110,177 @@ impl Coordinator {
+            SystemVars::max_replicas_per_cluster,
+            "cluster replica",
+            MAX_REPLICAS_PER_CLUSTER.name(),
+        )?;
```
Removed.

antiguru commented on src/adapter/src/coord/sequencer/cluster.rs:
```diff
@@ -322,6 +526,421 @@ impl Coordinator {
+            SystemVars::max_replicas_per_cluster,
+            "cluster replica",
+            MAX_REPLICAS_PER_CLUSTER.name(),
+        )?;
```
Removed.

tr0njavolta commented on doc/user/content/sql/create-cluster.md:
```diff
@@ -55,6 +55,27 @@ Clusters containing sources and sinks can have at most one replica.
+## Managed and unmanaged clusters
+
+Materialize offers managed and unmanaged clusters.
+Managed clusters offer a simple interface that only requires the user to specify a common size and replication factor, while Materialize manages the associated cluster replicas.
+Unmanaged clusters allow the user to specify the cluster replicas manually, offering more flexibility at the expense of more complexity.
+While both can be interchangeably used to host computations, they differ in how they are configured.
```
```suggestion
Managed clusters reduce the amount of manual configuration needed to create replicas. Managed clusters require the user to specify a common size and replication factor, while Materialize manages the associated cluster replicas.
Unmanaged clusters allow the user to specify the cluster replicas manually, offering more flexibility at the expense of more complexity.
While both can be interchangeably used to host computations, they differ in how they are configured.
```

tr0njavolta reviewed (commented):
Some suggestions for language and placement and questions about the feature.

benesch commented:
> > Item No 5. I think this can use the default cluster size setting rather than throwing an error:
> 
> I don't want to introduce a default cluster size as part of this PR, but would prefer to do this in follow-up work.

Agree! In fact I don't think we should introduce a default size until we have autoscaling clusters, at which point the default cluster can be one with a minimum and maximum size that work for a lot of use cases. Right now there is no good default cluster size to choose because workloads are so varied.

benesch commented:
Just took this for a spin. This is amazing! The ergonomics of this feel even better in practice than they did in theory.

The only thing I spotted is that the error messages don't offer the helpful hints mentioned in the design doc. Happy for this to be follow up work but we should file an issue so we don't forget:

```
materialize=> create cluster replica foo.managed_r3 size 'small';
ERROR:  cannot modify managed cluster foo

materialize=> alter cluster bar set (size = 'small');
ERROR:  SIZE not supported for unmanaged clusters
```

My one remaining point of minor stress is that this work is not feature flagged. It will be instantly stable. Are we happy enough with the UX that we're willing to commit to its stability? I am, personally, but I'm biased because I had such a significant hand in the design doc.

antiguru commented:
I filed #20000 (:tada:) to improve the error messages and hints.

antiguru commented:
> > > So we expect the replicas to follow a certain naming convention so that they can be converted to managed?
> > 
> > 
> > Yes. It's something that we reserve the right to change in the future, and I'm wondering whether we should simply use the `r1`, `r2`, ... pattern.
> 
> I'm wondering the same thing. The `managed_...` naming is a bit awkward when you want to convert an unmanaged cluster to a managed cluster. You have to type `ALTER CLUSTER REPLICA c.my_replica RENAME TO managed_r1`, and having an unmanaged cluster replica named `managed_r1`, even if you're never meant to stay in that state for long, seems quite confusing!

I changed the naming pattern to `r1`, `r2`, ... because we don't get any benefit out of using different names, only that it adds complexity.

jkosh44 reviewed (approved):
The feature gate commit LGTM
